STM32 DMAU5 DRIVER : add support for  Linked Lists  in DMA driver

djiatsaf-st opened the pull request:
This PR will add support for continuous transfer in the DMAU5 driver.

Some STM32 series, like STM32U5, STM32H5, and STM32WBA, use linked lists with the GPDMA peripheral to achieve continuous transfer. 
A sample driver has been updated to demonstrate how to use this feature for UART communication with the STM32WBA series

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -274,7 +274,9 @@ static void dma_stm32_irq_handler(const struct device *dev, uint32_t id)
 	} else if (stm32_dma_is_tc_irq_active(dma, id)) {
 		/* Assuming not cyclic transfer */
-		stream->busy = false;
+		if (stream->cyclic == false) {
+			stream->busy = false;
+		}
```
Not strictly linked to LLDMA. Please make it a dedicated commit with proper justification.

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -500,18 +505,38 @@ static int dma_stm32_configure(const struct device *dev,
 	DMA_InitStruct.Request = config->dma_slot;
 
-	LL_DMA_Init(dma, dma_stm32_id_to_stream(id), &DMA_InitStruct);
+	if (config->head_block->source_reload_en == 0) {
+		LL_DMA_Init(dma, dma_stm32_id_to_stream(id), &DMA_InitStruct);
```
`LL_DMA_Init` now depends on `if (config->head_block->source_reload_en == 0)`.
Is that safe not doing it if not the case ?

djiatsaf-st replied:
Good point,
From my point of view, this function must be initialized only in the non-cyclic mode with its corresponding parameters.

Then, after testing if I could remove this dependency and let the `LL_DMA_Init()` function run outside, I realized that there was a malfunction with a loss of the first characters of the first data frame in the console, thereafter, the rest of the data frames are well received throughout the communication.
This behavior could be due to the fact that during this initialization phase, some parameters of DMA_InitStruct are not compatible in the mode configured above.

So for me it is safe to put this dependency.

erwango replied:
I'll need to dig a bit on that matter, but if this change is required please make it a dedicated commit with proper justification.

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -500,18 +505,38 @@ static int dma_stm32_configure(const struct device *dev,
+		LL_DMA_Init(dma, dma_stm32_id_to_stream(id), &DMA_InitStruct);
+	} else { /* cyclic transfer */
+		/* Setting GPDMA request */
+		NodeConfig.SrcIncMode = DMA_InitStruct.SrcIncMode;
+		NodeConfig.DestIncMode = DMA_InitStruct.DestIncMode;
+		NodeConfig.SrcDataWidth = DMA_InitStruct.SrcDataWidth;
+		NodeConfig.DestDataWidth = DMA_InitStruct.DestDataWidth;
+		NodeConfig.Direction = DMA_InitStruct.Direction;
+		NodeConfig.Request = DMA_InitStruct.Request;
+
+		/* Continuous transfers with Linked List */
+		stream->cyclic = true;
+		LL_DMA_CreateLinkNode(&NodeConfig, &Node_GPDMA_Channel);
+		LL_DMA_ConnectLinkNode(&Node_GPDMA_Channel, LL_DMA_CLLR_OFFSET5,
+				       &Node_GPDMA_Channel, LL_DMA_CLLR_OFFSET5);
+		LL_DMA_SetLinkedListBaseAddr(dma, dma_stm32_id_to_stream(id),
+					     (uint32_t)&Node_GPDMA_Channel);
+		LL_DMA_List_Init(dma, dma_stm32_id_to_stream(id), &DMA_InitLinkedListStruct);
+		LL_DMA_ConfigLinkUpdate(dma, dma_stm32_id_to_stream(id),
+					(LL_DMA_UPDATE_CTR1 | LL_DMA_UPDATE_CTR2 |
+					 LL_DMA_UPDATE_CBR1 | LL_DMA_UPDATE_CSAR |
+					 LL_DMA_UPDATE_CDAR | LL_DMA_UPDATE_CLLR),
+					(uint32_t)&Node_GPDMA_Channel);
```
Should be protected by DT_HAS_COMPAT_STATUS_OKAY(st_stm32u5_dma)

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -349,10 +351,20 @@ static int dma_stm32_configure(const struct device *dev,
+
+	LL_DMA_NodeStructInit(&NodeConfig);
+	LL_DMA_ListStructInit(&DMA_InitLinkedListStruct);
+
```
Remove this line

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -349,10 +351,20 @@ static int dma_stm32_configure(const struct device *dev,
 	LL_DMA_StructInit(&DMA_InitStruct);
 
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32u5_dma)
+	/*  Linked list Node  and structure initialization */
+	LL_DMA_InitNodeTypeDef NodeConfig;
+	static LL_DMA_LinkNodeTypeDef Node_GPDMA_Channel;
+	LL_DMA_InitLinkedListTypeDef DMA_InitLinkedListStruct;
```
Declarations should be done at the begining of a code block (before `LL_DMA_StructInit(&DMA_InitStruct);` in this case)

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -349,10 +351,20 @@ static int dma_stm32_configure(const struct device *dev,
+	LL_DMA_ListStructInit(&DMA_InitLinkedListStruct);
+
+#endif
+	int ret;
```
No reason to move this line

erwango reviewed (changes requested):
Some comments, but otherwise general direction looks good to me.

djiatsaf-st committed:
drivers: dma: cyclic dma handling

keeping the stream busy when cyclic DMA is used.

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>

djiatsaf-st committed:
drivers: dma: add linked list in dmau5 driver

- Init linked list node and structure
- configure source and destination address
- Setting gpdma request mode
- Configure linked list node and structure for
continous transfer, enable half transfert irq.

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>
co-authored-by: Lubos Koudelka <lubos.koudelka@st.com>

djiatsaf-st committed:
samples: boards: st: uart: circular_dma: update circular dma sample

Add the nucleo_wba55cg board, which implements continuous transfer
with GPDMA and a linked list.

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>
